dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/validation/hooks/AbstractTrackerValidationHook.java

```diff
@@ -0,0 +1,393 @@
+            {
+                errorReporter.addError( newReport( TrackerErrorCode.E1011 )
+                    .withObject( te )
+                    .addArg( te.getOrgUnit() ) );
+            }
+
+        }
+        else if ( bundle.getImportStrategy().isUpdate() )
+        {
+            TrackedEntityInstance tei = PreheatHelper.getTrackedEntityInstance( bundle, te.getTrackedEntity() );
+
+            if ( tei.getOrganisationUnit() == null )
+            {
+                errorReporter.addError( newReport( TrackerErrorCode.E1011 )
+                    .withObject( te )
+                    .addArg( tei ) );
+            }
+        }
+    }
+
+    protected boolean textPatternValueIsValid( TrackedEntityAttribute attribute, String value, String oldValue )
+    {
+        return Objects.equals( value, oldValue ) ||
+            TextPatternValidationUtils.validateTextPatternValue( attribute.getTextPattern(), value ) ||
+            reservedValueService.isReserved( attribute.getTextPattern(), value );
+    }
+
+    private boolean validateTextPattern( ValidationErrorReporter errorReporter, Attribute attr,
+        TrackedEntityAttribute teAttr,
+        TrackedEntityAttributeValue teiAttributeValue )
+    {
+        if ( teAttr.isGenerated() && teAttr.getTextPattern() != null )
+        //&& ??? !importOptions.isSkipPatternValidation()
+        // MortenO: How should we deal with this in the new importer?
+        {
+            String oldValue = teiAttributeValue != null ? teiAttributeValue.getValue() : null;
+            if ( !textPatternValueIsValid( teAttr, attr.getValue(), oldValue ) )
+            {
+                errorReporter.addError( newReport( TrackerErrorCode.E1008 )
+                    .addArg( attr.getValue() ) );
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    private boolean validateFileNotAlreadyAssigned( ValidationErrorReporter errorReporter, Attribute attr,
+        TrackedEntityInstance tei )
+    {
+        boolean attrIsFile = attr.getValueType() != null && attr.getValueType().isFile();
+
+        if ( attrIsFile && tei != null )
+        {
+            List<String> existingValues = new ArrayList<>();
+
+            tei.getTrackedEntityAttributeValues().stream()
+                .filter( attrVal -> attrVal.getAttribute().getValueType().isFile() )
+                .forEach( attrVal -> existingValues.add( attrVal.getValue() ) );
+
+            FileResource fileResource = fileResourceService.getFileResource( attr.getValue() );
+            if ( fileResource != null && fileResource.isAssigned() && !existingValues.contains( attr.getValue() ) )
+            {
+                errorReporter.addError( newReport( TrackerErrorCode.E1009 )
+                    .addArg( attr.getValue() ) );
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    private boolean validateAttrValueType( ValidationErrorReporter errorReporter, Attribute attr,
```

Comment:
try not to use so many abbreviations, you can use `attribute` `trackedEntityAttributeService` etc